src/RewardsStreamerMP.sol

```diff
@@ -99,6 +100,60 @@ contract RewardsStreamerMP is IStakeManager, TrustedCodehashAccess, ReentrancyGu
         account.lastMPUpdateTime = block.timestamp;
     }
 
+    function lock(uint256 lockPeriod) external onlyTrustedCodehash nonReentrant {
+        if (lockPeriod == 0) {
+            revert StakingManager__InvalidLockingPeriod();
+        }
+
+        Account storage account = accounts[msg.sender];
+        if (account.stakedBalance == 0) {
+            revert StakingManager__InsufficientBalance();
+        }
+
+        if (account.lockUntil > 0) {
+            revert StakingManager__AlreadyLocked();
+        }
+
+        _updateGlobalState();
+        _updateAccountMP(msg.sender);
+
+        // Calculate remaining lock time if any
+        uint256 remainingLockTime = 0;
+        if (account.lockUntil > block.timestamp) {
+            remainingLockTime = account.lockUntil - block.timestamp;
+        }
+
+        // Calculate new total lock time
+        uint256 newTotalLockTime = remainingLockTime + lockPeriod;
+        if (newTotalLockTime > MAX_LOCKUP_PERIOD) {
+            revert StakingManager__InvalidLockingPeriod();
+        }
+
+        // Calculate additional bonus MP based on the new lock period
+        uint256 currentLockMultiplier =
+            remainingLockTime > 0 ? (remainingLockTime * MAX_MULTIPLIER * SCALE_FACTOR) / MAX_LOCKUP_PERIOD : 0;
+
+        uint256 newLockMultiplier = (newTotalLockTime * MAX_MULTIPLIER * SCALE_FACTOR) / MAX_LOCKUP_PERIOD;
+
+        // Calculate the differential in multiplier
+        uint256 multiplierDiff = newLockMultiplier - currentLockMultiplier;
```

Comment:
All the above of determining the `multiplierDiff` is not necessary if we agree to not allow for locking after a stake amount has already locked.

Had a discussion with @gravityblast about this. Basically we need to decide if vaults are allowed to lock multiple times after stake. If we do want to allow it, we also need to keep track of the total amount of time they have staked.

The reason is that, otherwise, vaults could:

- Stake and lock for say 1 year
- Wait until lock time has passed
- Lock again for 1 year
- Repeat.

While we would have a max cap on the `maxMP`, would also need a max cap on total lock up time or rather total time staked to the prevent users from locking up funds without gaining more MP.

We currently also don't allow for staking again when there's also stake in the vault, so to simplify things we can just not lock locking at all if it was already locked (no matter how long).

Thoughts @3esmit ?

Reply:
I would prefer to avoid locking multiple time, it makes it simpler and it's similar to the `stake` function.